File: 60060_가사검색.py
class Node:
    def __init__(self):
        self.nodeDict = {}
        self.lenDict = {}
        
originalRoot = Node()
reverseRoot = Node()
def updateTrie(root, myStr, reverse=False):
    node = root
    lenStr = len(myStr)
    if reverse == True:
        myStr = reversed(myStr)
    for i, c in enumerate(myStr):
        key = lenStr-i
        node.lenDict[key] = node.lenDict.setdefault(key, 0) + 1

        if c in node.nodeDict:
            node = node.nodeDict[c]
            
        else:
            node.nodeDict[c] = Node()
            node = node.nodeDict[c]
    return

def findEndNode(root, myStr):
    node = root
    for c in myStr:
        if c in node.nodeDict:
            node = node.nodeDict[c]
        else:
            return None
    return node

def getCount(query):
    s = query.split("?")
    if query[0] == '?':
        node = findEndNode(reverseRoot, reversed(s[-1]))
    else:
        node = findEndNode(originalRoot, s[0])
    if node == None:
        return 0 # trie에 없는 케이스
    cnt = node.lenDict.setdefault(query.count("?"),0)
    return cnt


def solution(words, queries):
    for word in words:
        updateTrie(originalRoot, word)
        updateTrie(reverseRoot, word, True)
    answer = [getCount(q) for q in queries]
    return answer

File: 60060_가사검색.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

class Node
{
public:
    Node* nextNode[26];
    map<int, int> lenMap;
};

Node forwardRoot;
Node reverseRoot;

void updateTrie(string word)
{
    Node* node = &forwardRoot;
    for (int i = 0; i < word.size(); i++)
    {
        int c = word[i] - 'a';
        node->lenMap[word.size()-i] += 1;
        if (node->nextNode[c] == 0)
        {
            node->nextNode[c] = new Node();
            node = node->nextNode[c];
        }
        else
        {
            node = node->nextNode[c];
;       }
    }
}
void updateReverseTrie(string word)
{
    Node* node = &reverseRoot;
    for (int i = word.size()-1; i >= 0; i--)
    {
        int c = word[i] - 'a';
        node->lenMap[i+1] += 1;
        if (node->nextNode[c] == 0)
        {
            node->nextNode[c] = new Node();
            node = node->nextNode[c];
        }
        else
        {
            node = node->nextNode[c];
        }
    }
}

Node* findEndNode(string word)
{
    Node* node = &forwardRoot;
    for (int i = 0; i < word.size(); i++)
    {
        int c = word[i] - 'a';
        if (node->nextNode[c] != 0)
            node = node->nextNode[c];
        else
            return 0;
    }
    return node;
}


Node* findReverseEndNode(string word)
{
    Node* node = &reverseRoot;
    for (int i = word.size()-1; i >= 0; i--)
    {
        int c = word[i] - 'a';
        if (node->nextNode[c] != 0)
            node = node->nextNode[c];
        else
            return 0;
    }
    return node;
}


int getCount(string query)
{
    int i = query.find('?');
    Node* node;
    if (i == 0)
    {
        i = query.rfind('?');

        node = findReverseEndNode(query.substr(i + 1, query.size()));
    }
    else
    {
        node = findEndNode(query.substr(0, i));
    }
    if (node == 0)
        return 0;

    return node->lenMap[std::count(query.begin(), query.end(), '?')];
}
vector<int> solution(vector<string> words, vector<string> queries) {
    vector<int> answer;
    for (int i = 0; i < words.size(); i++)
    {
        updateTrie(words[i]);
        updateReverseTrie(words[i]);
    }

    for (int i = 0; i < queries.size(); i++)
    {
        answer.push_back(getCount(queries[i]));
    }
    return answer;
}




int main()
{
    vector<string> words = { "frodo", "front", "frost", "frozen", "frame", "kakao" };
    solution(words, { "fro??", "????o", "fr???", "fro???", "pro?" });
    return 0;
}